Refactor heading block to share more code with web version.

SergioEstevao opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

This PR refactors the heading block to share the edit function with the web version. On the process, it adds the align button to the header block in mobile.
~It also reduces the number of headers available on the block toolbar in order to make the align buttons visible without the need for scrolling.~

~The only issue that I have is the way it shows the extra headers level on the Block setting modal~

~@iamthomasbishop  Should we convert this to another type of setting, that looks better on mobile?~

We now have an implementation of the dropdown menu in mobile that uses a bottom sheet to show the options available.

<img src="https://user-images.githubusercontent.com/651601/77238389-5a8bd980-6bc7-11ea-9fc9-b91868826dd8.gif" width=225 />

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

This can be tested on GB-mobile using this PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1990
## Screenshots <!-- if applicable -->

<img width="300" alt="Screenshot 2020-03-09 at 21 10 13" src="https://user-images.githubusercontent.com/651601/76258524-a778c300-624b-11ea-98ea-1bf51a5196df.png">


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

SergioEstevao committed:
Refactor heading block to share more code with web version.

youknowriad commented on packages/block-library/src/heading/edit.js:
```diff
@@ -54,17 +54,19 @@ function HeadingEdit( {
 					onChange={ ( newLevel ) =>
 						setAttributes( { level: newLevel } )
 					}
+					isCollapsed={ Platform.OS === 'web' }
```
Do mobile toolbars support isCollapsed prop? Wht happens if we keep it true for mobile too?

SergioEstevao replied:
They don't show., it looks collapsed is equivalent to hidden in mobile.

youknowriad replied:
Do you think we should update the toolbar implementation to just ignore that prop instead?

SergioEstevao replied:
I can update the code for the native ToolbarGroupCollapsed to simple do the same as the non-collapsed version, until we get a proper implementation of a collapsed toolbar for mobile.

youknowriad commented on packages/block-library/src/heading/edit.js:
```diff
@@ -81,7 +83,9 @@ function HeadingEdit( {
 					identifier="content"
-					tagName={ Block[ tagName ] }
+					tagName={
+						Platform.OS === 'web' ? Block[ tagName ] : tagName
```
I think `Block` component should have an equivalent in native. This is going to become a central component to write blocks. Ideally all new blocks use it. cc @ellatrix

ellatrix replied:
Yes, ideally native has a `Block` component too. I'm not sure what it will take there. I don't know enough about the native side, but happy to collaborate with someone.

SergioEstevao replied:
I'm working on to have an equivalent ( more a mock at this stage) on this PR: https://github.com/WordPress/gutenberg/pull/20772

Will update this PR when the above is merged in.

iamthomasbishop commented:
> Should we convert this to another type of setting, that looks better on mobile?

@sergioestevao I know the web extracts the heading levels in their entirety into the inspector, but I’ve always felt this was odd, especially when pitted as-is to mobile. I think these should all be inside the block toolbar, and while I understand there is reasonable rationale for not surfacing H1s, I don’t love the idea of losing H5 and H6 in the toolbar. 

Maybe this is a good opportunity to consider tackling collapsible groups for things like this (also for alignment controls, etc). I remember having some earlier discussions on this issue, but can’t seem to find that discussion atm. I’ll try to find it tomorrow and lay out a few rough sketches as well to demonstrate what the potential paths are.

SergioEstevao commented:
> I don’t love the idea of losing H5 and H6 in the toolbar.

@iamthomasbishop One advantage of not having those is that it makes the alignment buttons visible on the toolbar, without need to scroll.

I think we have a global problem in the toolbar when we have more commands that are not in the screen, I don't think our users know that they can scroll horizontally on the toolbar to get more commands.

SergioEstevao committed:
Merge branch 'master' into rnmobile/add_align_to_heading_block

SergioEstevao committed:
Remove branching for RN.

We now have an implementation of Block for RN.

iamthomasbishop commented:
> One advantage of not having those is that it makes the alignment buttons visible on the toolbar, without need to scroll.

> I think we have a global problem in the toolbar when we have more commands that are not in the screen, I don't think our users know that they can scroll horizontally on the toolbar to get more commands.

Both good points, although I think the scrolling concern is a more complex discussion that we should have separately. For now, implementing collapsible groups would mitigate the bulk of the cases where there is overflow so we should probably start there.

SergioEstevao committed:
Remove platform checks at block level.

SergioEstevao commented:
> For now, implementing collapsible groups would mitigate the bulk of the cases where there is overflow so we should probably start there.

I think this will be a broader work, and should be in a separate PR of this one. Meanwhile you prefer to have all header level in the toolbar and the block settings removed for mobile?

iamthomasbishop commented:
> I think this will be a broader work, and should be in a separate PR of this one.

@SergioEstevao Agree if we do further exploration, but before we push that work aside, let me propose what could be a nice temporary solution 😄 If it will require much additional effort, let me know and we will proceed with the solution mentioned below.

What do you think about doing something _extremely_ simple like tap-to-cycle-headings in the toolbar? The heading buttons would be collapsed into a single button, and tapping it would cycle through heading levels. In theory, we could do the same thing with alignment, etc. Here's a quick and dirty prototype for example:

<a href="https://user-images.githubusercontent.com/1200432/76451671-c4251000-639d-11ea-8ad1-96590f9166c5.gif"><img src="https://user-images.githubusercontent.com/1200432/76451671-c4251000-639d-11ea-8ad1-96590f9166c5.gif" width="375"></a>

--------

If you don't think the above would be much effort, for now, let's consider doing one of the following (in no particular order):

* Keep toolbar as-is, showing H2-H6 (takes up lots of space)
* Show all heading levels in the toolbar (takes up slightly more space)
* Only show H2, H3, and H4 in the toolbar like web (takes up less space, but requires us to add a control in the settings sheet — not ideal for the user and additional work right now)
* If we wanted to go _super_ simple, we could even show an ActionSheet w/ the heading levels

Let me know if you have any preferences or questions @SergioEstevao.

SergioEstevao committed:
Fix lint-errors.

SergioEstevao commented:
> Only show H2, H3, and H4 in the toolbar like web (takes up less space, but requires us to add a control in the settings sheet — not ideal for the user and additional work right now)

@iamthomasbishop at the moment I have the above implemented, so it's exactly like the web.

I can try to give it a go on the rotating buttons idea, but I have a feeling that it will very confusing for our users...

iamthomasbishop commented:
> I can try to give it a go on the rotating buttons idea, but I have a feeling that it will very confusing for our users...

As a temporary measure (temp. because we will be establishing a more holistic solution at some point), what do you think about relying on a native component like the ActionSheet (on Android, we would use a BottomSheet)? Example:

<a href="https://user-images.githubusercontent.com/1200432/76783904-b944f380-6780-11ea-8cfc-2b88a76942d5.gif"><img src="https://user-images.githubusercontent.com/1200432/76783904-b944f380-6780-11ea-8cfc-2b88a76942d5.gif" width="375"></a>

I think this would be pretty intuitive for a user and allow us to apply the pattern to other groups, such as alignment, etc. in a way that doesn't seem very expensive.

SergioEstevao commented:
I gave it a go and implement the dropdown like this: 

![HeaderDropdown](https://user-images.githubusercontent.com/651601/77213230-0b2fa580-6b02-11ea-8f35-e4a2c5f7c52e.gif)

What do you say @iamthomasbishop ?

SergioEstevao committed:
Implement Dropdown menu for React Native.

SergioEstevao committed:
Increase headers icon size.

SergioEstevao committed:
Show all heading levels on the block controls in RN

We also hide the headings settings in RN because all of them are
available in the block controls.

SergioEstevao commented:
@iamthomasbishop I followed your feedback and here is the latest version:

<img src="https://user-images.githubusercontent.com/651601/77238389-5a8bd980-6bc7-11ea-9fc9-b91868826dd8.gif" width=325 />

I styled the dropdown menu like the block settings and I added the checkmark icon to show current option selected.

SergioEstevao committed:
Merge branch 'master' into rnmobile/add_align_to_heading_block

etoledom commented on packages/components/src/dropdown-menu/index.native.js:
```diff
@@ -1,5 +1,173 @@
+						onClose={ onClose }
+					>
+						{ isFunction( children ) ? children( props ) : null }
+						<PanelBody title={ label }>
```
Could we use the Picker component to reuse functionality? (packages/components/src/mobile/picker/)

SergioEstevao replied:
The Picker is using an ActionSheet in iOS that doesn't translate to the interface we want here.
Or are you saying to refactor the Picker component to be like like this visually?

etoledom replied:
Oh, I thought we wanted the ActionSheet on iOS based on @iamthomasbishop 's [comment](https://github.com/WordPress/gutenberg/pull/20745#issuecomment-599664171).

Still, as this layout is already in place somewhere else, maybe we could add a prop to the picker component, something like `useNativeActionSheet`, and use the code currently from the Android implementation if that prop is `false`. Anyway, not sure if this is the best way to go or if it will make things unnecessary complicated.

iamthomasbishop commented:
@SergioEstevao I think this is looking pretty solid, and if we do stay this route (BottomSheet), I have some suggestions in terms of design details to shore up. Per @etoledom's question re: actionSheets, I'm not opposed to staying with a more native component like ActionSheet on iOS, but I would assume we're going to need a BottomSheet for Android.

Feedback on BottomSheet as you've shown above:

* Left margin of table row separators should be inset so that the lines start at the text label, not icon (56 on iOS, 72 on Android, IIRC)
* Last table row shouldn’t have a separator-bottom
* What are the margin values on the left/right sides of the sheet contents? It seems slightly more (24px?) than our default 16px
* Can we make the `checkmark` icon our primary blue?
* Can we add a separator on left and right side of the heading level button on the block toolbar?
* Would you mind adding screenshots of what this looks like on Android after addressing the above notes?

SergioEstevao committed:
Style separators to have left margin.

SergioEstevao committed:
Color selected checkmark on cell

SergioEstevao committed:
Add separators for dropdown menus

SergioEstevao committed:
Remove extra padding on panel for dropdown

SergioEstevao commented:
@iamthomasbishop I addressed all the issues you commented above, here is the end result:

### Android 
| Toolbar | Dropdown |
|---------|-----------|
| ![screen-android-toolbar](https://user-images.githubusercontent.com/651601/77492531-b3ae7400-6e38-11ea-9ac2-3ae67416c623.png) | ![screen-android-dropdown](https://user-images.githubusercontent.com/651601/77492536-b5783780-6e38-11ea-8d29-70cbcbe43aac.png) |

### iOS light mode

| Toolbar | Dropdown |
|---------|-----------|
| ![screen-ios-light-toolbar](https://user-images.githubusercontent.com/651601/77492537-b6a96480-6e38-11ea-837d-48aa8bd8afda.png) | ![screen-ios-light-dropdown](https://user-images.githubusercontent.com/651601/77492543-b741fb00-6e38-11ea-9cf6-c288d37afdf7.png) |

### iOS dark mode

| Toolbar | Dropdown |
|---------|-----------|
| ![screen-ios-dark-toolbar](https://user-images.githubusercontent.com/651601/77492547-b8732800-6e38-11ea-850c-55ec072aadc4.png) | ![screen-ios-dark-dropdown](https://user-images.githubusercontent.com/651601/77492545-b7da9180-6e38-11ea-9b41-a48584d45add.png) |

SergioEstevao committed:
Merge branch 'master' into rnmobile/add_align_to_heading_block

Tug commented on packages/components/src/text/mixins.js:
```diff
@@ -1,11 +1,8 @@
  * Internal dependencies
  */
 import { fontFamily } from './font-family';
+import css from './emotion-css';
```
That does not seem scalable. Maybe we could export `emotion` from `@wordpress/element` or something like that?

SergioEstevao replied:
@gziolo do you think that `wordpress/element` will be the right place for this?

gziolo replied:
I think the idea is to keep it as a bundled dependency of `@wordpress/components` until we decide that it scales well. @youknowriad and @ItsJonQ know better. I didn't follow that recently.

Tug replied:
Just found out about [@emotion/primitives](https://www.npmjs.com/package/@emotion/primitives), we should probably use that

SergioEstevao replied:
@Tug and I talked a bit more and because emotion is only being used on the Text component we will keep the import here.

Tug commented on packages/components/src/dropdown-menu/index.native.js:
```diff
@@ -1,5 +1,182 @@
+				);
+			} }
+			renderContent={ ( { isOpen, onClose, ...props } ) => {
+				return (
```
It seems what differs from the web is mostly this renderContent function. Maybe we could use a common version of DropdownMenu for both web and mobile and create 2 different versions of a new component `content.js` and `content.native.js`

SergioEstevao replied:
We will keep this change for a follow-up PR when we a more clear design direction for drowpdown menu looks in RN.

Tug reviewed (approved):
Nice work @SergioEstevao ! I had a couple thoughts, but otherwise I believe it's good to go 👍

gziolo commented on packages/components/src/index.native.js:
```diff
@@ -46,6 +46,7 @@ export { default as withFocusOutside } from './higher-order/with-focus-outside';
 export { default as withFocusReturn } from './higher-order/with-focus-return';
 export { default as withNotices } from './higher-order/with-notices';
 export { default as withSpokenMessages } from './higher-order/with-spoken-messages';
+export * from './text';
```
@ItsJonQ something to align with your other PR where you change the way it's exported for web.

ItsJonQ replied:
Ah yes! Thanks for the heads up.

I'm doing something similar (or related) in this PR:
https://github.com/WordPress/gutenberg/pull/21088

iamthomasbishop commented:
@SergioEstevao Looking good, just a couple minor requests to wrap up, and I've provided a visual example below:

* Remove separators on Android
* iOS text labels should sit at the 56px keyline (see example below)
* The text labels on the rows seem to have an additional inset of ~4-8px. Can we remove this? ([example](https://cloudup.com/cUy6OMM8OPy))
* I think we can make the section title more concise. Let's change `Change heading level` to `Heading level`.

<a href="https://user-images.githubusercontent.com/1200432/77550403-1591cc80-6e7f-11ea-8a3e-eaff325e78f1.png"><img src="https://user-images.githubusercontent.com/1200432/77550403-1591cc80-6e7f-11ea-8a3e-eaff325e78f1.png" width="700"></a>

Also likely not related to this PR, but if possible on Dark Mode, can we map the heading icons on the sheet and in the toolbar to the neutral gray? // cc @etoledom because I think you did some work on this recently, maybe you have a suggestion.

SergioEstevao commented:
@iamthomasbishop here is a screenshot with the updated iOS and Android versions:
| iOS | Android |
|-----|---------|
| ![Simulator Screen Shot - iPhone 11 Pro Max - 2020-03-25 at 17 18 51](https://user-images.githubusercontent.com/651601/77566305-5a3c5880-6ebd-11ea-8e33-da181c54bfac.png) | ![screen](https://user-images.githubusercontent.com/651601/77566646-de8edb80-6ebd-11ea-9d03-78a9c8407c0d.png) |

SergioEstevao committed:
Fix separator on iOS and Android.

iamthomasbishop commented:
Looks solid! Ship it! 🚀

youknowriad commented on packages/block-library/src/heading/edit.js:
```diff
@@ -97,6 +99,7 @@ function HeadingEdit( { attributes, setAttributes, mergeBlocks, onReplace } ) {
 						[ `has-text-align-${ align }` ]: align,
 					} ) }
 					placeholder={ placeholder || __( 'Write heading…' ) }
+					textAlign={ align }
```
This is the only thing that's bothering me, in this PR. It's redundunt with the style prop in the web and it can easily get removed since it's not used at all in the web version.

SergioEstevao replied:
This is a recurrent issue we have on the mobile implementations. We don't have access to CSS so we end up recurring to extra props or container classes to do the work that CSS does. I'm all open to suggestions on how can we avoid these issues.
@mtias told me there is some approach being discussed about Global Styling approach using props for the components that could be a way to make the implementation less dependent of CSS styling.
